File: task3/main.py
def find_reachable(graph, start):
    visited = set()

    def dfs(node):
        visited.add(node)
        for neighbor in graph[node]:
            if neighbor not in visited:
                dfs(neighbor)

    dfs(start)
    visited.discard(start)
    return visited

# Вариант с файлом:
# with open("input.txt") as f:
#     lines = f.read().splitlines()
#     n, m = map(int, lines[0].split())
#     graph = {i: [] for i in range(n)}
#     for line in lines[1:m+1]:
#         u, v = map(int, line.split())
#         graph[u].append(v)
#     start = int(lines[m+1])

n, m = map(int, input("Введите количество вершин и рёбер: ").split())
graph = {i: [] for i in range(n)}
for _ in range(m):
    u, v = map(int, input("Введите ребро (u v): ").split())
    graph[u].append(v)  # ориентированный — только в одну сторону

start = int(input("Начальная вершина: "))
reachable = find_reachable(graph, start)
print("Достижимые вершины:", reachable)

File: task3/main.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

void dfs(vector<vector<int>>& graph, int node, vector<bool>& visited) {
    visited[node] = true;
    for (int neighbor : graph[node]) {
        if (!visited[neighbor]) {
            dfs(graph, neighbor, visited);
        }
    }
}

int main() {
    int n, m;
    cout << "Введите количество вершин и рёбер: ";
    cin >> n >> m;

    // Вариант с файлом:
    // ifstream f("input.txt");
    // f >> n >> m;
    // vector<vector<int>> graph(n);
    // for (int i = 0; i < m; i++) {
    //     int u, v; f >> u >> v;
    //     graph[u].push_back(v); // ориентированный
    // }
    // int start; f >> start;

    vector<vector<int>> graph(n);
    for (int i = 0; i < m; i++) {
        int u, v;
        cout << "Введите ребро (u v): ";
        cin >> u >> v;
        graph[u].push_back(v); // ориентированный — только в одну сторону
    }

    int start;
    cout << "Начальная вершина: ";
    cin >> start;

    vector<bool> visited(n, false);
    dfs(graph, start, visited);

    cout << "Достижимые вершины из " << start << ": ";
    for (int i = 0; i < n; i++) {
        if (visited[i] && i != start)
            cout << i << " ";
    }

    return 0;
}